src/main/scala/org/graphframes/lib/AggregateMessages.scala

```diff
@@ -0,0 +1,161 @@
+ *    triplet
+ *  - [[AggregateMessages.Builder.sendToDst()]] sends a message to the destination vertex of each
+ *    triplet
+ *  - [[AggregateMessages.Builder.agg()]] specifies an aggregation function for aggregating the
+ *    messages sent to each vertex.  It also runs the aggregation, computing a DataFrame
+ *    with one row for each vertex which receives > 0 messages.  The DataFrame has 2 columns:
+ *     - vertex column ID (named [[GraphFrame.ID]])
+ *     - aggregate from messages sent to vertex (with the name given to the [[Column]] specified
+ *       in [[AggregateMessages.Builder.agg()]])
+ *
+ * When specifying the messages and aggregation function, the user may reference columns using:
+ *  - [[AggregateMessages.src]]: column for source vertex of edge
+ *  - [[AggregateMessages.edge]]: column for edge
+ *  - [[AggregateMessages.dst]]: column for destination vertex of edge
+ *  - [[AggregateMessages.msg]]: message sent to vertex (for aggregation function)
+ *
+ * Note: If you use this operation to write an iterative algorithm, you may want to use
+ * [[AggregateMessages.getCachedDataFrame()]] as a workaround for caching issues.
+ *
+ * @example We can use this function to compute the in-degree of each vertex
+ * {{{
+ * val g: GraphFrame = Graph.textFile("twittergraph")
+ * val inDeg: DataFrame =
+ *   g.aggregateMessages().sendToDst(lit(1)).agg(sum(AggregateMessages.msg))
+ * }}}
+ */
+object AggregateMessages extends Logging with Serializable {
+
+  class Builder private[graphframes] (g: GraphFrame)
+    extends Arguments with Serializable {
+
+    private var msgToSrc: Option[Column] = None
+
+    /** Send message to source vertex */
+    def sendToSrc(value: Column): this.type = {
+      msgToSrc = Some(value)
+      this
+    }
+
+    /** Send message to source vertex, specifying SQL expression as a String */
+    def sendToSrc(value: String): this.type = sendToSrc(expr(value))
+
+    private var msgToDst: Option[Column] = None
+
+    /** Send message to destination vertex */
+    def sendToDst(value: Column): this.type = {
+      msgToDst = Some(value)
+      this
+    }
+
+    /** Send message to destination vertex, specifying SQL expression as a String */
+    def sendToDst(value: String): this.type = sendToDst(expr(value))
+
+    /**
+     * Run the aggregation, returning the resulting DataFrame of aggregated messages.
+     * This is a lazy operation, so the DataFrame will not be materialized until an action is
+     * executed on it.
+     *
+     * This returns a DataFrame with schema:
+     *  - column "id": vertex ID
+     *  - aggCol: aggregate result
+     * If you need to join this with the original [[GraphFrame.vertices]], you can run an inner
+     * join of the form:
+     * {{{
+     *   val g: GraphFrame = ...
+     *   val aggResult = g.aggregateMessages.sendToSrc(msg).agg(aggFunc)
+     *   aggResult.join(g.vertices, ID)
+     * }}}
+     */
+    def agg(aggCol: Column): DataFrame = {
```

Comment:
I'm not sure we'll be able to give a better error message than Spark SQL can, unless we really pick apart the user's Column to identify references.  I'm also worried about efficiency since the query analysis can get pretty hairy.  I'd prefer to leave this as is for now.  We can add checks later on if needed.